### ISSUE TITLE
False Negative: Redundant null checks are not tracked through aliasing and control-flow variants.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5642]
# False Negative: Redundant null checks are not tracked through aliasing and control-flow variants.

Version
2.48.0

## Checker
- Checker id: `RedundantNullCheck`
- Checker description: Detects redundant null checks on expressions that are statically determined to be non-null based on language semantics or nullness annotations.

## Description of the false negative
`diff/output` marks 8 false negatives for `RedundantNullCheck`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are string reference equality.

### `PosCase2.java`
Source: `output/codeql/UselessNullCheck/src/main/java/scensct/core/pos/PosCase2.java`

The core operation is a reference-equality check involving strings. The misuse is still a plain reference comparison on string values, which should not depend on whether one operand is a literal or a variable. That is also how the benchmark classifies the sample: A final field initialized with a non-null value directly compared to null should be flagged as useless null-check.

```java
// A final field initialized with a non-null value directly compared to null should be flagged as useless null-check.
package scensct.core.pos;

public class PosCase2 {
    private final String NAME = "constant"; // Final field with non-null initializer
    public void check() {
        if (NAME != null) { // Useless null-check on known non-null final field // [REPORTED LINE]
            System.out.println("always true");
        }
    }
}
```

### `PosCase3.java`
Source: `output/codeql/UselessNullCheck/src/main/java/scensct/core/pos/PosCase3.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A final field initialized with a non-null value passed to Objects.isNull() should be flagged as useless null-check.

```java
// A final field initialized with a non-null value passed to Objects.isNull() should be flagged as useless null-check.
package scensct.core.pos;

import java.util.Objects;

public class PosCase3 {
    private final String NAME = "constant"; // Final field with non-null initializer
    public void check() {
        if (Objects.isNull(NAME)) { // Useless null-check via utility method // [REPORTED LINE]
            System.out.println("impossible");
        }
    }
}
```

### `PosCase4.java`
Source: `output/codeql/UselessNullCheck/src/main/java/scensct/core/pos/PosCase4.java`

The core operation is a reference-equality check involving strings. The misuse is still a plain reference comparison on string values, which should not depend on whether one operand is a literal or a variable. That is also how the benchmark classifies the sample: A cast expression where the inner expression is known non-null checked for null should be flagged as useless null-check.

```java
// A cast expression where the inner expression is known non-null checked for null should be flagged as useless null-check.
package scensct.core.pos;

public class PosCase4 {
    public static void main(String[] args) {
        Object obj = "string"; // Known non-null reference
        String casted = (String) obj; // Cast expression
        if (casted == null) { // Useless null-check on cast of known non-null // [REPORTED LINE]
            System.out.println("impossible");
        }
    }
}
```

### `PosCase5.java`
Source: `output/codeql/UselessNullCheck/src/main/java/scensct/core/pos/PosCase5.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A cast expression where the inner expression is known non-null passed to Objects.nonNull() should be flagged as useless null-check.

```java
// A cast expression where the inner expression is known non-null passed to Objects.nonNull() should be flagged as useless null-check.
package scensct.core.pos;

import java.util.List;
import java.util.Objects;

public class PosCase5 {
    public static void main(String[] args) {
        Object obj = "string"; // Known non-null reference
        List casted = (List) obj; // Cast expression (assume List is imported, using raw type for minimalism)
        if (Objects.nonNull(casted)) { // Useless null-check via utility method // [REPORTED LINE]
            System.out.println("always true");
        }
    }
}
```

### `PosCase6.java`
Source: `output/codeql/UselessNullCheck/src/main/java/scensct/core/pos/PosCase6.java`

The core operation is a reference-equality check involving strings. The misuse is still a plain reference comparison on string values, which should not depend on whether one operand is a literal or a variable. That is also how the benchmark classifies the sample: A ternary conditional expression with both branches non-null directly compared to null should be flagged as useless null-check.

```java
// A ternary conditional expression with both branches non-null directly compared to null should be flagged as useless null-check.
package scensct.core.pos;

public class PosCase6 {
    public static void main(String[] args) {
        boolean flag = true;
        String result = flag ? "yes" : "no"; // Both branches are non-null literals
        if (result == null) { // Useless null-check on ternary with non-null branches // [REPORTED LINE]
            System.out.println("impossible");
        }
    }
}
```

### `PosCase7.java`
Source: `output/codeql/UselessNullCheck/src/main/java/scensct/core/pos/PosCase7.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A ternary conditional expression with both branches non-null passed to Objects.nonNull() should be flagged as useless null-check.

```java
// A ternary conditional expression with both branches non-null passed to Objects.nonNull() should be flagged as useless null-check.
package scensct.core.pos;

import java.util.Objects;

public class PosCase7 {
    public static void main(String[] args) {
        boolean cond = false;
        String result = cond ? "a" : "b"; // Both branches are non-null literals
        if (Objects.nonNull(result)) { // Useless null-check via utility method // [REPORTED LINE]
            System.out.println("always true");
        }
    }
}
```

### `PosCase8.java`
Source: `output/codeql/UselessNullCheck/src/main/java/scensct/core/pos/PosCase8.java`

The core operation is a reference-equality check involving strings. The misuse is still a plain reference comparison on string values, which should not depend on whether one operand is a literal or a variable. That is also how the benchmark classifies the sample: Inside a branch guarded by a null-check, the same variable directly compared to null should be flagged as useless null-check.

```java
// Inside a branch guarded by a null-check, the same variable directly compared to null should be flagged as useless null-check.
package scensct.core.pos;

public class PosCase8 {
    public static void main(String[] args) {
        String s = args.length > 0 ? args[0] : null;
        if (s != null) { // Guarding null-check
            if (s == null) { // Useless null-check inside guarded branch // [REPORTED LINE]
                System.out.println("impossible");
            }
        }
    }
}
```

### `PosCase9.java`
Source: `output/codeql/UselessNullCheck/src/main/java/scensct/core/pos/PosCase9.java`

The core operation is a reference-equality check involving strings. The misuse is still a plain reference comparison on string values, which should not depend on whether one operand is a literal or a variable. That is also how the benchmark classifies the sample: Inside a branch guarded by a null-check, the same variable passed to Objects.nonNull() should be flagged as useless null-check.

```java
// Inside a branch guarded by a null-check, the same variable passed to Objects.nonNull() should be flagged as useless null-check.
package scensct.core.pos;

import java.util.Objects;

public class PosCase9 {
    public static void main(String[] args) {
        Object obj = args.length > 0 ? args[0] : null;
        if (obj != null) { // Guarding null-check
            if (Objects.nonNull(obj)) { // Useless null-check via utility method inside guarded branch // [REPORTED LINE]
                System.out.println("always true");
            }
        }
    }
}
```

## Cause analysis
The rule is documented as follows: Detects redundant null checks on expressions that are statically determined to be non-null based on language semantics or nullness annotations.

In this set the problem shows up around string reference equality. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `RedundantNullCheck`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.
